src/consensus/pbft_manager.cpp

```diff
@@ -1367,18 +1370,76 @@ bool PbftManager::comparePbftBlockScheduleWithDAGblocks_(blk_hash_t const &pbft_
   if (!pbft_block) {
     return false;
   }
-
-  return comparePbftBlockScheduleWithDAGblocks_(*pbft_block).second;
+  return comparePbftBlockScheduleWithDAGblocks_(pbft_block).has_value();
 }
 
-std::pair<vec_blk_t, bool> PbftManager::comparePbftBlockScheduleWithDAGblocks_(PbftBlock const &pbft_block) {
-  auto const &anchor_hash = pbft_block.getPivotDagBlockHash();
+std::optional<vec_blk_t> PbftManager::comparePbftBlockScheduleWithDAGblocks_(std::shared_ptr<PbftBlock> pbft_block) {
+  // If cert sync block is already populated with this pbft block, no need to do verification again
+  if (cert_sync_block_.pbft_blk && cert_sync_block_.pbft_blk->getBlockHash() == pbft_block->getBlockHash()) {
+    vec_blk_t dag_blocks_order;
+    dag_blocks_order.reserve(cert_sync_block_.dag_blocks.size());
+    std::transform(cert_sync_block_.dag_blocks.begin(), cert_sync_block_.dag_blocks.end(),
+                   std::back_inserter(dag_blocks_order), [](const DagBlock &dag_block) { return dag_block.getHash(); });
+    return std::move(dag_blocks_order);
+  }
+  auto const &anchor_hash = pbft_block->getPivotDagBlockHash();
   auto dag_blocks_order = dag_mgr_->getDagBlockOrder(anchor_hash).second;
   if (!dag_blocks_order.empty()) {
-    return std::make_pair(std::move(dag_blocks_order), true);
+    std::unordered_set<trx_hash_t> trx_set;
+    std::vector<trx_hash_t> transactions_to_query;
+    DbStorage::MultiGetQuery db_query(db_);
+    db_query.append(DbStorage::Columns::dag_blocks, dag_blocks_order);
```

Comment:
isn't is better to get them from dag_blk_mgr_->getDagBlock? I think we can be 99,9% sure those block are still cached at this time

Reply:
yes, and I added the change for the other two comments